src/Database/LSMTree/Internal.hs

```diff
@@ -254,39 +303,71 @@ closeSession Session{sessionState} =
+      -- the table's fields (see 'withOpenTable'). We use more fine-grained
+      -- synchronisation for various mutable parts of an open table.
+      --
       -- TODO: RWVars only work in IO, not in IOSim.
       --
       -- TODO: how fair is an RWVar?
       tableHandleState :: RWVar (TableHandleState m h)
     }
 
+-- | A table handle may assume that its corresponding session is still open as
+-- long as the table handle is open. A session's global resources, and therefore
+-- resources that are inherited by the table, will only be released once the
+-- session is sure that no tables are open anymore.
 data TableHandleState m h =
     TableHandleOpen {-# UNPACK #-} !(TableHandleEnv m h)
   | TableHandleClosed
 
 data TableHandleEnv m h = TableHandleEnv {
-    -- === Inherited from Session
-    tableSessionRoot :: !FsPath
-  , tableHasFS       :: !(HasFS m h)
-  , tableHasBlockIO  :: !(HasBlockIO m h)
+    -- === Session
+    tableSessionRoot         :: !FsPath
+  , tableHasFS               :: !(HasFS m h)
+  , tableHasBlockIO          :: !(HasBlockIO m h)
+    -- | Open tables are tracked in the corresponding session, so when a table
+    -- is closed it should become untracked (forgotten).
+  , tableSessionUntrackTable :: !(m ())
     -- === Table-specific
-  , tableConfig      :: !TableConfig
-  , tableWriteBuffer :: !(StrictMVar m (WriteBuffer))
+    --
+    -- TODO: more fine-grained concurrency for table-specific mutable state.
+  , tableConfig              :: !TableConfig
+  , tableWriteBuffer         :: !(StrictMVar m (WriteBuffer))
     -- | A hierarchy of levels. The vector indexes double as level numbers.
-  , tableLevels      :: !(StrictMVar m (V.Vector (Level (Handle h))))
+  , tableLevels              :: !(StrictMVar m (V.Vector (Level (Handle h))))
     -- | Cache of flattened 'levels'.
     --
     -- INVARIANT: when 'level's is modified, this cache should be updated as
     -- well, for example using 'mkLevelsCache'.
-  , tableCache       :: !(StrictMVar m (LevelsCache (Handle h)))
+  , tableCache               :: !(StrictMVar m (LevelsCache (Handle h)))
   }
 
+-- | 'withOpenTable' ensures that the table stays open for the duration of the
+-- provided continuation.
+--
+-- NOTE: any operation except 'close' can use this function.
+{-# SPECIALISE withOpenTable :: TableHandle IO HandleIO -> (TableHandleEnv IO HandleIO -> IO a) -> IO a #-}
+withOpenTable ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => TableHandle m h
+  -> (TableHandleEnv m h -> m a)
+  -> m a
+withOpenTable th action = RW.with (tableHandleState th) $ \case
+    TableHandleClosed -> throwIO ErrTableClosed
+    TableHandleOpen thEnv -> action thEnv
+
+{-# SPECIALISE tableIsClosed :: TableHandle IO HandleIO -> IO Bool #-}
+tableIsClosed ::
+     m ~ IO
+  => TableHandle m h
+  -> m Bool
+tableIsClosed th = RW.with (tableHandleState th) $ \case
+    TableHandleClosed -> pure True
+    TableHandleOpen _ -> pure False
```

Comment:
Oh yeah, same here. Might just want to inline instead of specialising.

Also, I'm not sure where `tableIsClosed` is supposed to be used, but it seems prone to race conditions. Maybe we want to only add it once we need it (and know whether it's safe to be used in that context)?

Reply:
I agree, `tableIsClosed` isn't used yet and is inherently racy. Lets add it on demand, i.e. not now.

Reply:
I don't really remember why I added it :P If it turns out I'll need it, I'll add it then